### ISSUE TITLE
show stat -1 2 -1 json  fails

### ISSUE THREAD

**Issue by gabrieltz** [haproxy/haproxy#1964]
### Detailed Description of the Problem

I have some tools that use the admin socket for stats and modifying settings. 
However i noticed that when i issue `show stat -1 2 -1 json` to get a backend list i only got handful ( i have ~40 backend on that server ) and the last piece of the json output was:
```json 
[{"errorStr":"output buffer too short"}]
```

I removed all frontends and started removing backends one by one to find the problematic one and ofcourse it was the last one 
`listen stats`.... 

I looked up the error string in the code and found multiple locations with `goto err;` that after adding a lot of printfs led me to this line:
 https://github.com/haproxy/haproxy/blob/4a04cd35ae89bf6a3bb7620f7a49804de3240ac4/src/stats.c#L792

so in stats_emit_json_data_field function where it returned false here: https://github.com/haproxy/haproxy/blob/4a04cd35ae89bf6a3bb7620f7a49804de3240ac4/src/stats.c#L483
because `old_len` and `out->data` where equal .
output after adding a printf before return 
```
.
.
.
old len : 14579  out->data: 14612
old len : 14782  out->data: 14815
old len : 14989  out->data: 15022
old len : 15189  out->data: 15222
old len : 15385  out->data: 15418
old len : 15588  out->data: 15621
old len : 15787  out->data: 15820
old len : 15981  out->data: 16014
old len : 16176  out->data: 16209
old len : 16378  out->data: 16378
```

##### THIS ONLY HAPPENS WHEN COMPILED WITH SSL / QUIC

* breaks `make TARGET=linux-glibc USE_OPENSSL=1 USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_SPLICE=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1 CPU=native USE_THREAD=1 USE_CPU_AFFINITY=1 V=1 -j6`
* breaks `make TARGET=linux-glibc USE_OPENSSL=1 USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_SPLICE=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1 V=1 -j6`
* breaks `make TARGET=linux-glibc USE_OPENSSL=1 USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_PROMEX=1 V=1 -j6`
* breaks `make TARGET=linux-glibc USE_OPENSSL=1 USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 V=1 -j6`
* works `make TARGET=linux-glibc USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 V=1 -j6`  ( no ssl ) 

### Expected Behavior

stats should always work 

### Steps to Reproduce the Behavior

1. use the minimal haproxy.cfg below
2. compile haproxy with one of the make recipies above ( with ssl ) 
3. run `haproxy -f debug-haproxy.cfg -d`
4. echo "show stat stats 2 -1 json" | socat /run/haproxy/admin-debug.sock stdio

### Do you have any idea what may have caused this?

My knowledge in C is limited to running gdb and adding printfs 

### What is your configuration?

```haproxy
the minimum config to reproduce this issue is 

global
	daemon
	maxconn 100
	log /dev/log local6
	chroot /var/lib/haproxy
	pidfile /run/haproxy/haproxy-debug.pid
	stats socket /run/haproxy/admin-debug.sock mode 660 level admin expose-fd listeners
	stats timeout 30s
	user haproxy
	group haproxy

defaults
        mode http
        timeout connect 4s
        timeout http-request 6s
        timeout client 60s
        timeout server 30s
	timeout http-keep-alive 3s
	backlog 262144

	option http-ignore-probes
	option http-server-close
        option dontlognull
	option log-separate-errors


        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /dev/null
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http


listen stats
	bind 127.0.0.1:8089
	mode http
	stats enable
	http-request use-service prometheus-exporter if { path /metrics }
	stats realm Haproxy\ Statistics
	stats uri /
	stats auth user:pass
```
```


### Output of `haproxy -vv`

```plain
/usr/sbin/haproxy -vv
HAProxy version 2.8-dev0-393e4e-60 2022/12/12 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.4.0-51-generic #56-Ubuntu SMP Mon Oct 5 14:28:49 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = native
  CC      = cc
  CFLAGS  = -O2 -march=native -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_EPOLL=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_THREAD=1 USE_LINUX_SPLICE=1 USE_OPENSSL=1 USE_ZLIB=1 USE_CPU_AFFINITY=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -PTHREAD_EMULATION +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL -OPENSSL_WOLFSSL -LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -51DEGREES_V4 -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT +QUIC +PROMEX -MEMORY_PROFILING +SHM_OPEN

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=12).
Built with OpenSSL version : OpenSSL 1.1.1s+quic  1 Nov 2022
Running on OpenSSL version : OpenSSL 1.1.1s+quic  1 Nov 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

```
-rwxr-xr-x 1 root root 18867856 Sep  9  2021 /usr/sbin/haproxy-2.4.4
-rwxr-xr-x 1 root root 18921504 Nov 18  2021 /usr/sbin/haproxy-2.4.8
-rwxr-xr-x 1 root root 19038568 Mar  8  2022 /usr/sbin/haproxy-2.4.14
-rwxr-xr-x 1 root root 19075480 May 24  2022 /usr/sbin/haproxy-2.4.17
-rwxr-xr-x 1 root root 24102008 Oct 12 14:53 /usr/sbin/haproxy-2.7-dev7
-rwxr-xr-x 1 root root 23585456 Nov 22 09:53 /usr/sbin/haproxy-2.7-dev9-c21a18-14
-rwxr-xr-x 1 root root 23552256 Nov 23 09:09 /usr/sbin/haproxy-2.7-dev9-92c2de-31
-rwxr-xr-x 1 root root 22651048 Nov 30 08:58 /usr/sbin/haproxy-2.7-dev10-4ede46-33  <== works fine up to this point
-rwxr-xr-x 1 root root 22832984 Dec 12 14:28 /usr/sbin/haproxy-2.8.0-dev0
lrwxrwxrwx 1 root root       28 Dec 12 14:28 /usr/sbin/haproxy -> /usr/sbin/haproxy-2.8.0-dev0
```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by Darlelet** [comment]
Some stats were added since 2.7-dev10:

- https://github.com/haproxy/haproxy/commit/745ce8e8ad8150569f368ab4663adda77a43c857
- https://github.com/haproxy/haproxy/commit/e06e31ea3b62ef8ccb911ac3969ae70f7bbb7574
- https://github.com/haproxy/haproxy/commit/7d6644e689f15b329789a355ea2812ea0223fe4f

Maybe you were flirting with output buffer space limit and the extra metrics that were introduced are enough to reveal the issue?
(16k is the default size for general purpose buffers in haproxy, according to your tests logs, your json output seems to reach max buffer size)

**Comment by Darlelet** [comment]
`Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200
`

Could you try increasing `tune.bufsize` to something slightly larger than 16k and see if it restores original behavior?

**Comment by gabrieltz** [comment]
You are right , increasing the bufsize to 20k works for the specific listen block. However the behaviour below is very strange
* 16k buffer size
```
# echo "show stat stats -1 -1 json" | socat /run/haproxy/admin.sock stdio 
[{"errorStr":"output buffer too short"}]
```

* 20k buffer size 
```
# echo "show stat stats -1 -1 json" | socat /run/haproxy/admin-debug.sock stdio | wc -c
16659
```

* 16k buffer again on another backend  ( this does not fail while being well over 16k ) 
```
# echo "show stat resizer-cluster -1 -1 json" | socat /run/haproxy/admin.sock stdio | wc -c
134638
```

The output is attached below
[resizer-stats.txt](https://github.com/haproxy/haproxy/files/10216578/resizer-stats.txt)

This seems very odd to me, but I'll use 20k buffer size for now 
thank you

**Comment by Darlelet** [comment]
> 16k buffer again on another backend ( this does not fail while being well over 16k )

My guess is that the complete json dump could be the result of multiple buffer aggregations depending on the proxy type and options. (Independent chunks <= 16k that are combined together and result in a much larger output/response)
And here json generation for stats proxy might be the only one running into the 16k default buffer limit for your setup.

But you're right, if you run in the issue with minimal config, perhaps this should be addressed on haproxy side because we keep adding new metrics over the time and this will get worse.

**Comment by gabrieltz** [comment]
I did another experiment with this very lean config file

```
global
        daemon
        maxconn 100
        log /dev/log local6
        chroot /var/lib/haproxy
        pidfile /run/haproxy/haproxy-debug.pid
        stats socket /run/haproxy/admin-debug.sock mode 660 level admin expose-fd listeners
        stats timeout 30s
        user haproxy
        group haproxy
        # tune.bufsize 20480 ### we have 16k ( the default )

defaults
        mode http

backend meatballs
        mode http
        balance first
        server hello 192.168.1.1:80 check
```

If i run show stat for the specific backend the buffer is too small :)
```
# echo "show stat meatballs 2 -1 json" | socat /run/haproxy/admin-debug.sock stdio
[{"errorStr":"output buffer too short"}]
```

but if i run `show stat json` the output is there and fits nicely in the buffer:
```
# echo "show stat json" | socat /run/haproxy/admin-debug.sock stdio | tail -c 100
tags":{"origin":"Metric","nature":"Counter","scope":"Process"},"value":{"type":"u64","value":0}}]]
```

So i don't think it's the buffer.

If i comment out this if and goto here, rebuild 
https://github.com/haproxy/haproxy/blob/4a04cd35ae89bf6a3bb7620f7a49804de3240ac4/src/stats.c#L792

and run the 1st command it works perfectly fine and the json output is ~13k , smaller than the bufsize.

```
2,"id":0,"field":{"pos":125,"name":"h1_bytes_out"},"processNum":1,"tags":{"origin":"Metric","nature":"Counter","scope":"Process"}},{"objType":"Backend","proxyId":2,"id":0,"field":{"pos":126,"name":"h1_spliced_bytes_in"},"processNum":1,"tags":{"origin":"Metric","nature":"Counter","scope":"Process"}},{"objType":"Backend","proxyId":2,"id":0,"field":{"pos":127,"name":"h1_spliced_bytes_out"},"processNum":1,"tags":{"origin":"Metric","nature":"Counter","scope":"Process"}}]]

root@blade5:~/haproxy# echo "show stat meatballs 2 -1 json" | socat /run/haproxy/admin-debug.sock stdio | wc -c
13871
```

I hope this helps

**Comment by Darlelet** [comment]
Thank you for the extra infos.
Json output is heavier than other output types and now we know for sure that we are beginning to reach the 16k default buffer limit with simple configuration as shown by your tests.

This lazy fix should make json stats generation future-proof:
```diff
diff --git a/include/haproxy/stats-t.h b/include/haproxy/stats-t.h
index 1a7d3d157..f63e7860f 100644
--- a/include/haproxy/stats-t.h
+++ b/include/haproxy/stats-t.h
@@ -539,6 +539,7 @@ struct show_stat_ctx {
 	int scope_str;		/* limit scope to a frontend/backend substring */
 	int scope_len;		/* length of the string above in the buffer */
 	int px_st;		/* STAT_PX_ST* */
+	int px_it;              /* data transmission iterator for multi-part aware dumping functions */
 	unsigned int flags;	/* STAT_* from stats-t.h */
 	int iid, type, sid;	/* proxy id, type and service id if bounding of stats is enabled */
 	int st_code;		/* the status code returned by an action */
diff --git a/src/stats.c b/src/stats.c
index 5fd4d3f96..c9dfe3caf 100644
--- a/src/stats.c
+++ b/src/stats.c
@@ -744,40 +744,40 @@ static void stats_print_rslv_field_json(struct buffer *out,
 
 /* Dump all fields from <stats> into <out> using a typed "field:desc:type:value" format */
 static int stats_dump_fields_json(struct buffer *out,
-                                  const struct field *stats, size_t stats_count,
-                                  unsigned int flags,
-                                  enum stats_domain domain)
+                                  const struct field *stats,
+                                  size_t stats_count,
+                                  struct show_stat_ctx *ctx)
 {
-	int field;
-	int started = 0;
+	int started = (ctx->px_it) ? 1 : 0;
+	int ready_data = 0;
 
-	if ((flags & STAT_STARTED) && !chunk_strcat(out, ","))
+	if (!ctx->px_it && (ctx->flags & STAT_STARTED) && !chunk_strcat(out, ","))
 		return 0;
-	if (!chunk_strcat(out, "["))
+	if (!ctx->px_it && !chunk_strcat(out, "["))
 		return 0;
 
-	for (field = 0; field < stats_count; field++) {
+	for (; ctx->px_it < stats_count; ctx->px_it++) {
 		int old_len;
+		int field = ctx->px_it;
 
 		if (!stats[field].type)
 			continue;
-
 		if (started && !chunk_strcat(out, ","))
 			goto err;
 		started = 1;
 
 		old_len = out->data;
-		if (domain == STATS_DOMAIN_PROXY) {
+		if (ctx->domain == STATS_DOMAIN_PROXY) {
 			stats_print_proxy_field_json(out, &stats[field],
-			                             stat_f[domain][field].name,
+			                             stat_f[ctx->domain][field].name,
 			                             field,
 			                             stats[ST_F_TYPE].u.u32,
 			                             stats[ST_F_IID].u.u32,
 			                             stats[ST_F_SID].u.u32,
 			                             stats[ST_F_PID].u.u32);
-		} else if (domain == STATS_DOMAIN_RESOLVERS) {
+		} else if (ctx->domain == STATS_DOMAIN_RESOLVERS) {
 			stats_print_rslv_field_json(out, &stats[field],
-			                            stat_f[domain][field].name,
+			                            stat_f[ctx->domain][field].name,
 			                            field);
 		}
 
@@ -792,19 +792,27 @@ static int stats_dump_fields_json(struct buffer *out,
 
 		if (!chunk_strcat(out, "}"))
 			goto err;
+		ready_data = out->data;
 	}
 
 	if (!chunk_strcat(out, "]"))
 		goto err;
 
+	ctx->px_it = 0; /* we're done */
 	return 1;
 
 err:
-	chunk_reset(out);
-	if (flags & STAT_STARTED)
-		chunk_strcat(out, ",");
-	chunk_appendf(out, "{\"errorStr\":\"output buffer too short\"}");
-	return 0;
+	if (!ready_data) {
+		/* not enough buffer space for a single entry.. */
+		chunk_reset(out);
+		if (ctx->flags & STAT_STARTED)
+			chunk_strcat(out, ",");
+		chunk_appendf(out, "{\"errorStr\":\"output buffer too short\"}");
+		return 0; /* hard error */
+	}
+	/* push ready data and wait for a new buffer to complete the dump */
+	out->data = ready_data;
+	return 1;
 }
 
 /* Dump all fields from <stats> into <out> using the HTML format. A column is
@@ -1628,7 +1636,7 @@ int stats_dump_one_line(const struct field *stats, size_t stats_count,
 	else if (ctx->flags & STAT_FMT_TYPED)
 		ret = stats_dump_fields_typed(&trash, stats, stats_count, ctx->flags, ctx->domain);
 	else if (ctx->flags & STAT_FMT_JSON)
-		ret = stats_dump_fields_json(&trash, stats, stats_count, ctx->flags, ctx->domain);
+		ret = stats_dump_fields_json(&trash, stats, stats_count, ctx);
 	else
 		ret = stats_dump_fields_csv(&trash, stats, stats_count, ctx->flags, ctx->domain);
 
@@ -3035,7 +3043,10 @@ int stats_dump_proxy_to_buffer(struct stconn *sc, struct htx *htx,
 	struct channel *rep = sc_ic(sc);
 	struct server *sv, *svs;	/* server and server-state, server-state=server or server->track */
 	struct listener *l;
+	int save_it;
 
+ more:
+	save_it = ctx->px_it;
 	chunk_reset(&trash);
 
 	switch (ctx->px_st) {
@@ -3098,6 +3109,9 @@ int stats_dump_proxy_to_buffer(struct stconn *sc, struct htx *htx,
 		if (stats_dump_fe_stats(sc, px)) {
 			if (!stats_putchk(rep, htx, &trash))
 				goto full;
+			if (ctx->px_it)
+				goto more;
+			save_it = 0;
 		}
 
 		ctx->obj2 = px->conf.listeners.n;
@@ -3132,6 +3146,9 @@ int stats_dump_proxy_to_buffer(struct stconn *sc, struct htx *htx,
 			if (stats_dump_li_stats(sc, px, l)) {
 				if (!stats_putchk(rep, htx, &trash))
 					goto full;
+				if (ctx->px_it)
+					goto more;
+				save_it = 0;
 			}
 		}
 
@@ -3205,6 +3222,9 @@ int stats_dump_proxy_to_buffer(struct stconn *sc, struct htx *htx,
 		if (stats_dump_be_stats(sc, px)) {
 			if (!stats_putchk(rep, htx, &trash))
 				goto full;
+			if (ctx->px_it)
+				goto more;
+			save_it = 0;
 		}
 
 		ctx->px_st = STAT_PX_ST_END;
@@ -3230,6 +3250,8 @@ int stats_dump_proxy_to_buffer(struct stconn *sc, struct htx *htx,
 
   full:
 	sc_need_room(sc);
+	/* restore previous it */
+	ctx->px_it = save_it;
 	return 0;
 }
 
@@ -3719,6 +3741,7 @@ static int stats_dump_proxies(struct stconn *sc,
 
 		ctx->obj1 = px->next;
 		ctx->px_st = STAT_PX_ST_INIT;
+		ctx->px_it = 0;
 	}
 
 	return 1;
@@ -3781,6 +3804,7 @@ static int stats_dump_stat_to_buffer(struct stconn *sc, struct htx *htx,
 			ctx->obj1 = proxies_list;
 
 		ctx->px_st = STAT_PX_ST_INIT;
+		ctx->px_it = 0;
 		ctx->state = STAT_STATE_LIST;
 		__fallthrough;
 
```
It enables `stats_dump_field_*` functions to make multipart (using multiple buffers) dumping, in case a single buffer is not big enough to hold the complete dump.
For now, only `stats_dump_fields_json()` makes use of this capability as it is by far the most verbose stats output type.
(csv, typed and html outputs should be good for a while and may use this capability if the need arises in some distant future)

**Comment by gabrieltz** [comment]
excellent , the patch works perfectly.

thank you

**Comment by wtarreau** [comment]
Just so that I'm sure I understood correctly, you mean that we've reached the point where dumping a *single* line of stats in the output takes more than 16kB in JSON ? I knew that was quite verbose (not as much as XML) but still that's a lot, it would mean we'd have several hundred indicators. That might be possible in fact, but I wouldn't have expected that many.

**Comment by Darlelet** [comment]
Not single lines, json stats dumping function is currently condensing multiple lines/metrics at once within a single trash buffer (per proxy).

**Comment by Darlelet** [comment]
Every `stats_dump_fields_*` functions are implemented this way:

```
for (field = 0; field < stats_count; ++field) {
   // chunk_append
}
```

And as each stat/metric takes more than 50 chars once outputted in json, the last 3 added metrics made us reach the limit for show stat json command with regular proxies.

show info json is based on the same logic. We still have some margin left but we're getting closer, I'm reaching 10k buffer size with `stats_dump_json_info_fields()` and very basic config files.

I'm proposing the following patches to address this limitation (currently targeting json as other output types are far less verbose, but it could be easily applied to csv, html and typed output as well):

[0001-BUG-MINOR-stats-fix-show-stat-json-buffer-limitation.patch.txt](https://github.com/haproxy/haproxy/files/10235861/0001-BUG-MINOR-stats-fix-show-stat-json-buffer-limitation.patch.txt)
[0002-MINOR-stats-make-show-info-json-future-proof.patch.txt](https://github.com/haproxy/haproxy/files/10235862/0002-MINOR-stats-make-show-info-json-future-proof.patch.txt)

Edit: `stats_dump_fields_html()` is not following the same logic, it is selectively picking specific metrics (not exhaustive), sorry for the impreciseness. (patching it would require more efforts, but hopefully it doesn't grow systematically with each added metric)

**Comment by wtarreau** [comment]
Thanks, you responded in your second comment to the question I started to ask :-) So in fact you're confirming that we're reaching 16kB with a single line. Hmmm looking at the output it's indeed particularly verbose, with all the field's full path being repeated for each entry. I understand better how we can reach this situation.

There are two things I'm still not fond of with your patch.

The first one is your iterator name "px_it" (or "part_it" in the latest patch) that does not evocate anything to me. Since you're using it to store the field number being dumped, just call it "field" and indicate that it's the current field being dumped. There's indeed nothing specific with this one, all other ones are iterators as well and the sole purpose of the ctx is to store them to restart from the same point. This one is just the field number where others are a server number or proxy number.

The second one is that it took us a while to finally arrive to the point where the 3 output functions had the exact same API and could be used interchangeably, and by doing this we're making a significant step backwards. What I would suggest instead is to have a first patch to just change the funtions' API to take the context in argument and that they all retrieve ->flags and ->domain from that context. A second patch to introduce the field number in the structure and the ability for the calling function to interrupt/restart in the middle of the line like you've done, and a final 3rd patch would enable this for the json output function. This way if we later face a problem, the bisect will be more accurate, and if we later need to implement this for the CSV or HTML output, we'd just restart from the 3rd patch and would know what to do exactly.

What do you think ?

**Comment by capflam** [comment]
I'm wondering if it is pertinent to dump all stats  in json, even those with no meanings for a scope. In the CSV output it is required to keep everything aligned. But in json, we can filter the fields per scope, just like in html. This will reduce the size for a given line.

**Comment by Darlelet** [comment]
> So in fact you're confirming that we're reaching 16kB with a single line.

Yes sorry, by "line" I was thinking about a single metric dump, not a full stat line, my mistake!

> The second one is that it took us a while to finally arrive to the point where the 3 output functions had the exact same API and could be used interchangeably, and by doing this we're making a significant step backwards. What I would suggest instead is to have a first patch to just change the funtions' API to take the context in argument and that they all retrieve ->flags and ->domain from that context. A second patch to introduce the field number in the structure and the ability for the calling function to interrupt/restart in the middle of the line like you've done, and a final 3rd patch would enable this for the json output function. This way if we later face a problem, the bisect will be more accurate, and if we later need to implement this for the CSV or HTML output, we'd just restart from the 3rd patch and would know what to do exactly.

You're right, this totally makes sense to me.

> I'm wondering if it is pertinent to dump all stats in json, even those with no meanings for a scope. In the CSV output it is required to keep everything aligned. But in json, we can filter the fields per scope, just like in html. This will reduce the size for a given line.

I agree with you @capflam, but this would introduce some noticeable difference in the way json dumps appear to the user, maybe this could be addressed later as part of a new release? (considering this behavior has not changed since it was added in 05ee213f8b27855d0c6578ced12127e2e1196266, maybe some scripts / parsers are "incorrectly" relying on it?)

**Comment by wtarreau** [comment]
> considering this behavior has not changed since it was added in https://github.com/haproxy/haproxy/commit/05ee213f8b27855d0c6578ced12127e2e1196266, maybe some scripts / parsers are "incorrectly" relying on it?)

I undeed have vague memories if some users who were using the json dump but were splitting it around field names using `cut` or `grep -o`, that might explain why we're still having the extremely verbose format. Another reason is that when we started doing this, there was no stored context beyond the few ints in the CLI's appctx, and there was no way to know what had changed since previous line. So I think that we're just dumping entire descriptions for each line and that the client reconstructs the whole table internally while consuming them. Now that we're having the stat_ctx it would become easier to know what level(s) we crossed (i.e. changed server or socket, changed proxy etc) and be able to terminate the previous dump by closing the required number of braces, and opening new ones with their description. This way we would eliminate the redundancy that can currently locate the object. However we'd still need an option to preserve the current format in case it's still in use (e.g "show stat json by-line"). We could also do that in two steps, one consisting in adding the new format as "compact" and the second change which would only affect the default output format (i.e. "compact" would become implicit). But that last one is a breaking change and would have to be done only in an odd version (e.g. 2.9).

**Comment by gabrieltz** [comment]
I must report that the buffer pool leak is back ( as discussed in #1903 with @a-denoyelle and @wtarreau ) .
I was running  2.8-dev0-f5994f-82 when it made a come-back. I don''t know if the above patch is to blame, as i added it on this version. 
The previous version used without leaking was 2.8-dev0-9192d2-54 and 2.8-dev0-4a04cd-66 ( on another instance ) 

![Screenshot from 2022-12-17 09-31-03](https://user-images.githubusercontent.com/13848662/208231112-ece26bbc-8c7c-44cc-a611-cabcb7b7c45f.png)

**Comment by wtarreau** [comment]
I would say the patch is unlikely to be responsible for this given that it only touches the stats, unless of course you're dumping stats very often like twice a second or so, which sounds unlikely. However what's strange is that the curve is very regular, which tends to indicate a result of a periodic activity.

Also in the range of commits you're mentioning, very few things were touched on the code. Mostly CI, regtests and coverity cleanups. There were also extra error checks in the H3 code. Oh, wait a minute, it looks like some of these could cause a mem leak on error, that was addressed later:
```
commit 788fc054016761d835cff5e560f9f250c4c739d6
Author: Amaury Denoyelle <adenoyelle@haproxy.com>
Date:   Thu Dec 15 10:53:55 2022 +0100

    BUG/MINOR: h3: fix memleak on HEADERS parsing failure
    
    If an error is triggered on H3 HEADERS parsing, the allocated buffer for
    HTX data is not freed.
```

So very likely there's an agent outside sending periodic (and invalid) requests to your site and causing such leaks. Most likely just updating to the latest master (46bea1c616 as of now) will address this! It also contains a fix for a rare case of crash affecting QUIC.

**Comment by gabrieltz** [comment]
There is no need to keep this issue open since it's fixed , thank you

**Comment by Darlelet** [comment]
Thanks for confirming Gabriel!

However we should keep the issue open until the fix is backported to 2.6. (we're using it as a reminder)
